hackathon-2024-03-watch-out

diegommm opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

[Add a brief description of what the feature or update does.]

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

DanCech committed:
initial naive implementation

DanCech committed:
Update pkg/services/store/entity/sqlstash/sql_storage_server.go

Co-authored-by: Igor Suleymanov <radiohead@users.noreply.github.com>

DanCech committed:
tidy up

DanCech committed:
add action column, batch watch events

DanCech committed:
initial implementation of broadcast-based watcher

DanCech committed:
fix up watch init

DanCech committed:
remove batching, it just adds needless complexity

DanCech committed:
use StreamWatcher

DanCech committed:
make broadcaster generic

DanCech committed:
add circular buffer to replay recent events to new watchers

DanCech committed:
loop within poll until all events are read

DanCech committed:
add index on entity_history.resource_version to support poller

DanCech committed:
increment r.Since when we send events to consumer

DanCech committed:
switch broadcaster and cache to use channels instead of mutexes

DanCech committed:
cleanup

diegommm committed:
initial commit of work

diegommm committed:
first version building, mostly complete, without tests

DanCech commented on pkg/services/store/entity/sqlstash/sql_storage_server.go:
```diff
@@ -1486,8 +1582,345 @@ func watchMatches(r *entity.EntityWatchRequest, result *entity.Entity) bool {
+
+	return nil
+}
+
```
there seems to be a lot of duplication between this code and the existing poll-based setup, is there a reason not to just update line ~147 to choose between the standard poller and the pg watch implementation, but keep the same broadcast code?

diegommm replied:
Actually, I think I can work around it to keep track of the postgres channels listened to separately and unlisten and terminate separately, and that would make the code more compatible with Broadcaster. I'll try to do that, but I'd like to wrap a bit first for the hackathon video as well. But I also think we may need the unbounded buffer channel implementation for the reasons above, wdyt?

Commit:
Fixes if/else so pgWatch is called when running postgres.

Initializes channel map using pgChannelName.

Removes channel close from cleanup func since it was closing the channel a 2nd time and panicking.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=83772) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=83772) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: DanCech<br/>:white_check_mark: diegommm<br/>:x: Owen Smallwood<br/><hr/>**Owen Smallwood** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=83772) it.</sub>

Commit:
Merge branch 'main' into hackathon-2024-03-watch-out

diegommm committed:
Fix listen not working

- added proper connection string
- added unbounded channels for notifications to prevent loss
- several other fixes and improvements

diegommm committed:
small improvements; close connection after no longer needed

diegommm committed:
fix condition

Commit:
Adds integration tests for watch. Still a WIP. Watch behaving non-deterministically.

owensmallwood commented on pkg/services/store/entity/tests/server_integration_test.go:
```diff
@@ -109,6 +112,84 @@ func requireVersionMatch(t *testing.T, obj *entity.Entity, m objectVersionMatche
+		require.NoError(t, err)
+
+		// watch client receives create
+		res, err := watchClient.Recv()
```
Sometimes this hangs forever... sometimes it doesn't 🤯

diegommm committed:
minor fixes in cleanup

diegommm committed:
Merge remote-tracking branch 'remotes/origin/hackathon-2024-03-watch-out-diego' into hackathon-2024-03-watch-out

diegommm committed:
remove initial SQL file

diegommm committed:
fix query

diegommm committed:
fix query

diegommm committed:
fix broker loop not terminating

Commit:
adds delete to test and some helper functions

Commit:
adds test

diegommm commented:
Closing as related work is now under development in https://github.com/grafana/grafana/pull/84629 and other PRs.
